internal/topology/variables/clusterclass_variable_validation_test.go

```diff
@@ -2593,6 +2593,109 @@ func Test_ValidateClusterClassVariable(t *testing.T) {
+						}},
+						AllOf: []clusterv1.JSONSchemaProps{{
+							XValidations: []clusterv1.ValidationRule{{
+								Rule: "should be forbidden",
+							}},
+						}},
+						OneOf: []clusterv1.JSONSchemaProps{{
+							XValidations: []clusterv1.ValidationRule{{
+								Rule: "should be forbidden",
+							}},
+						}},
+					},
+				},
+			},
+			wantErrs: []validationMatch{
+				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.allOf[0].x-kubernetes-validations"),
+				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.anyOf[0].x-kubernetes-validations"),
+				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.not.x-kubernetes-validations"),
+				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.oneOf[0].x-kubernetes-validations"),
+			},
+		},
+		{
+			name: "fail if defaults are under oneOf/anyOf/allOf/not",
+			clusterClassVariable: &clusterv1.ClusterClassVariable{
+				Name: "variableA",
+				Schema: clusterv1.VariableSchema{
+					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+						Type: "number",
+						Not: &clusterv1.JSONSchemaProps{
+							Default: &apiextensionsv1.JSON{
+								Raw: []byte(`42.0`),
+							},
+						},
+						AnyOf: []clusterv1.JSONSchemaProps{{
+							Default: &apiextensionsv1.JSON{
+								Raw: []byte(`42.0`),
+							},
+						}},
+						AllOf: []clusterv1.JSONSchemaProps{{
+							Default: &apiextensionsv1.JSON{
+								Raw: []byte(`42.0`),
+							},
+						}},
+						OneOf: []clusterv1.JSONSchemaProps{{
+							Default: &apiextensionsv1.JSON{
+								Raw: []byte(`42.0`),
+							},
+						}},
+					},
+				},
+			},
+			wantErrs: []validationMatch{
+				forbidden("must be undefined to be structural", "spec.variables[variableA].schema.openAPIV3Schema.allOf[0].default"),
+				forbidden("must be undefined to be structural", "spec.variables[variableA].schema.openAPIV3Schema.anyOf[0].default"),
+				forbidden("must be undefined to be structural", "spec.variables[variableA].schema.openAPIV3Schema.not.default"),
+				forbidden("must be undefined to be structural", "spec.variables[variableA].schema.openAPIV3Schema.oneOf[0].default"),
+			},
+		},
+		{
+			name: "pass if oneOf/anyOf/allOf/not schemas are valid",
+			clusterClassVariable: &clusterv1.ClusterClassVariable{
+				Name: "variableA",
+				Schema: clusterv1.VariableSchema{
+					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+						Type: "string",
+						Not: &clusterv1.JSONSchemaProps{
+							Format: "email",
+						},
+						AnyOf: []clusterv1.JSONSchemaProps{{
+							Format: "ipv4",
+						}, {
+							Format: "ipv6",
+						}},
+						AllOf: []clusterv1.JSONSchemaProps{{
+							Format: "ipv4",
+						}, {
+							Format: "ipv6",
+						}},
+						OneOf: []clusterv1.JSONSchemaProps{{
+							Format: "ipv4",
+						}, {
+							Format: "ipv6",
+						}},
+					},
+				},
+			},
+		},
```

Comment:
Let's add this test case:
```go
{
			name: "pass & fail correctly for int-or-string in oneOf/anyOf/allOf schemas",
			clusterClassVariable: &clusterv1.ClusterClassVariable{
				Name: "variableA",
				Schema: clusterv1.VariableSchema{
					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
						Type: "object",
						Properties: map[string]clusterv1.JSONSchemaProps{
							"anyOfExampleField": { // Only valid variant
								Type: "object",
								AnyOf: []clusterv1.JSONSchemaProps{{
									Type: "integer",
								}, {
									Type: "string",
								}},
							},
							"allOfExampleField": {
								Type: "object",
								AllOf: []clusterv1.JSONSchemaProps{{
									Type: "integer",
								}, {
									Type: "string",
								}},
							},
							"oneOfExampleField": {
								Type: "object",
								OneOf: []clusterv1.JSONSchemaProps{{
									Type: "integer",
								}, {
									Type: "string",
								}},
							},
						},
					},
				},
			},
			wantErrs: []validationMatch{
				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.properties[allOfExampleField].allOf[0].type"),
				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.properties[allOfExampleField].allOf[1].type"),
				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.properties[oneOfExampleField].oneOf[0].type"),
				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.properties[oneOfExampleField].oneOf[1].type"),
			},
		},
```

(I want to make sure int-or-string works)

Reply:
Added.